gazebo_ros/src/executor.cpp

```diff
@@ -34,14 +49,100 @@ Executor::~Executor()
   spin_thread_.join();
 }
 
-void Executor::run()
+void Executor::shutdown()
 {
-  spin();
+  rclcpp::shutdown();
 }
 
-void Executor::shutdown()
+void
+Executor::add_node(
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr, bool notify)
 {
-  rclcpp::shutdown();
+  std::lock_guard<std::mutex> wait_lock(wait_mutex_);
+  rclcpp::Executor::add_node(node_ptr, notify);
+}
+
+void
+Executor::add_node(std::shared_ptr<rclcpp::Node> node_ptr, bool notify)
+{
+  this->add_node(node_ptr->get_node_base_interface(), notify);
+}
+
+void
+Executor::remove_node(
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr, bool notify)
+{
+  std::lock_guard<std::mutex> wait_lock(wait_mutex_);
+  rclcpp::Executor::remove_node(node_ptr, notify);
+}
+
+void
+Executor::remove_node(std::shared_ptr<rclcpp::Node> node_ptr, bool notify)
+{
+  this->remove_node(node_ptr->get_node_base_interface(), notify);
+}
+
+void
+Executor::spin()
+{
+  if (spinning.exchange(true)) {
+    throw std::runtime_error("spin() called while already spinning");
+  }
+  RCLCPP_SCOPE_EXIT(this->spinning.store(false); );
+  std::vector<std::thread> threads;
+  size_t thread_id = 0;
+  {
+    for (; thread_id < number_of_threads_ - 1; ++thread_id) {
+      auto func = std::bind(&Executor::run, this);
+      threads.emplace_back(func);
+    }
+  }
+
+  run();
+  for (auto & thread : threads) {
+    thread.join();
+  }
+}
+
+void
+Executor::run()
+{
+  while (rclcpp::ok(this->context_) && spinning.load()) {
+    rclcpp::AnyExecutable any_exec;
+    std::lock_guard<std::mutex> wait_lock(wait_mutex_);
```

Comment:
Why not just using the rclcpp `SingleThreadedExecutor`?

I mean, this is just creating N threads that are running this loop in a mutually exclusive fashion, so there's actually zero benefit of having many threads (the threads are all the time fighting for the lock, and only one is doing work).

Reply:
Yeah, I think you're right. I overlooked that when I increased the scope of the lock (compared to the upstream implementation in [`rclcpp::MultiThreadedExecutor`](https://github.com/ros2/rclcpp/blob/adebda52c541533d878a4d0bae2ed112403c1e2a/rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp#L76-L114).

Reply:
I've restored the original scope of `wait_mutex_`, which I think keeps the multi-threaded nature of the executor. See f9c0b54.